### PR TITLE
option to colorize output

### PR DESCRIPTION
**Pull request by mohoff**
Donezo:

- [x] add helper `prettyColorizedSchema` to colorize the output (gray out lines starting with `#`, yellow for types, e.g. `<string>`, blue/red for optional/optional fields respectively.

No extra dependencies involved.

### CODE REVIEW

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 1]
```diff
@@ -38,8 +39,23 @@ prettyParserDoc = prettySchema . explainParser
 prettySchema :: Schema -> Text
 prettySchema = renderStrict . layoutPretty defaultLayoutOptions . schemaDoc
 
+-- | Render a schema as pretty and colorized text.
+--
+-- This is meant for humans.
+-- The output may look like YAML but it is not.
+prettyColorizedSchema :: Schema -> Text
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:49]: It's `gray` in the US but you used `grey`, so please also use `Colourise`.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 2]
```diff
@@ -57,29 +73,32 @@ emptyComments = Comments []
...
             Just t -> c <> comment t
-          e :: Doc () -> Comments -> Doc ()
+          e :: Doc Color -> Comments -> Doc Color
           e s cs' =
             case mkCommentsMDoc cs' of
               Nothing -> s
-              Just cd -> vsep [cd, s]
+              Just cd -> vsep [cd, annotate Yellow s]
        in \case
             EmptySchema -> e "# Nothing to parse" cs
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: This is also a comment I guess, so best to make it grey too.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 3]
```diff
@@ -57,29 +73,32 @@ emptyComments = Comments []
...
             Comments l -> Just $ align $ vsep $ map mkComment l
           addMComment :: Comments -> Maybe Text -> Comments
           addMComment c = \case
             Nothing -> c
             Just t -> c <> comment t
-          e :: Doc () -> Comments -> Doc ()
+          e :: Doc Color -> Comments -> Doc Color
           e s cs' =
             case mkCommentsMDoc cs' of
               Nothing -> s
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: You probably want to make `s` yellow here too.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 4]
```diff
@@ -38,8 +39,23 @@ prettyParserDoc = prettySchema . explainParser
...
 prettySchema = renderStrict . layoutPretty defaultLayoutOptions . schemaDoc
 
+-- | Render a schema as pretty and colorized text.
+--
+-- This is meant for humans.
+-- The output may look like YAML but it is not.
+prettyColorizedSchema :: Schema -> Text
+prettyColorizedSchema = renderSimplyDecorated id startColor resetColor . layoutPretty defaultLayoutOptions . schemaDoc
+
+startColor :: Color -> Text
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: You can put `startColor` and `resetColor` in a `where` clause of `prettyColorisedSchema` because you don't need it outside of that function.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 5]
```diff
@@ -57,29 +73,32 @@ emptyComments = Comments []
...
           ge = go emptyComments
-          mkComment :: Doc () -> Doc ()
-          mkComment = ("# " <>)
-          mkCommentsMDoc :: Comments -> Maybe (Doc ())
+          mkComment :: Doc Color -> Doc Color
+          mkComment = (annotate Grey) . ("# " <>)
+          mkCommentsMDoc :: Comments -> Maybe (Doc Color)
           mkCommentsMDoc = \case
             Comments [] -> Nothing
             Comments l -> Just $ align $ vsep $ map mkComment l
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: This is where the actual rendering of the comments happens.
I think we should keep the `Doc` inside the `Comments` constructor un-annotated, and then only annotate it here when we get them out of this `Comments` constructor.
That way definitely all comments have been annotated correctly and it's checked by the compiler.

**Review by NorfairKing — COMMENTED** [review]
Some minor comments.

I'd also be ok with colour-coding the `required` and `optional` keys (not the comments) as `red` and `blue` respectively but that's optional.

Could you also add an entry to the changelog and bump the version so that I can make a release out of this?

**Comment by NorfairKing** [comment]
intray
It seems to work when I use an incorrect config file, when I use `prettyColorizedSchema` in `YamlParse.Applicative.IO` (I'd like to make that change in this PR).
But it doesn't seem to work when I use it in in `confDescWith` in `YamlParse.Applicative.OptParse`.
I wish there was a way to get it to "just work" with the optparse stuff.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 1 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:49]: renamed to gray

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 2 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: done

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 3 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: done

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 4 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: ah right, I keep forgetting that

Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: done

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 5 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: Yeah I agree. I made some changes that should implement your suggestion

**Comment by mohoff** [comment]
@NorfairKing Fixed visual bugs for `FieldSchema`, `AltSchema`, and `EmptySchema`. Also changed the display of `ObjectSchema` to have type `<object>`. Added blue+red for optional+required. Addressed all comments so far.

I don't know about `confDescWith` or how to fix it.

**Comment by NorfairKing** [comment]
colour
It's really pretty!

* About `confDescWith`, see https://github.com/pcapriotti/optparse-applicative/issues/391. It's not something for this PR because we'd have to reimplement the pretty printer for optparse.
* While we're here. Could we make the keys bright white? Now the keys have the same colour as the comments for me.
* Could you also make this change please, then we have colours automatically if something fails to parse:

```
--- a/yamlparse-applicative/src/YamlParse/Applicative/IO.hs
+++ b/yamlparse-applicative/src/YamlParse/Applicative/IO.hs
@@ -49,7 +49,7 @@ readFirstConfigFile files = go files
                         fs -> "While parsing files:" : map (("* " <>) . toFilePath) fs
                       referenceMsgs =
                         [ "Reference: ",
-                          T.unpack $ prettySchema $ explainParser (yamlSchema :: YamlParser a)
+                          T.unpack $ prettyColorizedSchema $ explainParser (yamlSchema :: YamlParser a)
```

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 6]
```diff
@@ -58,32 +77,32 @@ comment :: Text -> Comments
...
-              Nothing -> s
-              Just cd -> vsep [cd, s]
+              Nothing -> annotate Yellow s
+              Just cd -> vsep [cd, annotate Yellow s]
        in \case
-            EmptySchema -> e "# Nothing to parse" cs
+            EmptySchema -> e emptyDoc $ addMComment cs $ Just "Nothing to parse"
             AnySchema -> e "<any>" cs
-            ExactSchema t -> e (pretty t) cs
+            ExactSchema t -> e (pretty t <+> annotate Gray "(exact)") cs
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: I like this change, but :

1. The colours don't work it seems:
colour_000
2. Could we put that in a comment instead of in parentheses so that colour-blind people don't get confused:
```
e (pretty t) (cs <> comment "exact")
```
We essentially never want to show information in colour only.

Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: Hm, for me it dims `(exact)` (blend yellow with gray) which i found to be looking nicely.

Even without colors, I find the distinction pretty clear so don't see an issue for colorblind people. But can change it for stylistic reasons if you like

Comment (NorfairKing) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: comment
How do you like this?

```
            ExactSchema t ->
              let mExactCommentDoc = mkCommentsMDoc $ comment "(exact)"
               in ( case mExactCommentDoc of
                      Nothing -> id
                      Just cd -> (<+> cd)
                  )
                    $ e (pretty t) cs
```
The blending you saw is likely not portable so we shouldn't rely on it.

**Comment by mohoff** [comment]
>     * While we're here. Could we make the keys bright white? Now the keys have the same colour as the comments for me.
> 

I think your color scheme is tweaked. For me, the keys are bright white and the comments grayed out. Before I try changes that I can't test, can you please either provide my a color code that reflects "bright white" in your terminal - or try yourself first?

**Comment by NorfairKing** [comment]
white
This is what I had in mind.

And here is my complete diff:
https://pastebin.com/prqUBVBd


The code you asked for is `"\x1b[37".

Please check what it looks like in your terminal too.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 7]
```diff
@@ -92,17 +111,21 @@ schemaDoc = go emptyComments
             ArraySchema t s -> "-" <+> align (go (addMComment cs t) s)
             -- The comments really only work on the object level
             -- so they are erased when going down
-            ObjectSchema t s -> e (ge s) (addMComment cs t)
+            ObjectSchema t s -> vsep
+              [ e "<object>" (addMComment cs t)
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:116]: Why did you add this piece? 
Now it looks like there are two pieces of an object expected:
object

**Comment by NorfairKing** [comment]
Also please make sure the ci passes by running `nix-shell` in the root and then running `pre-commit run -a`.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 6 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs]: Done using `fromJust`

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs [hunk 8]
```diff
@@ -127,7 +128,7 @@ schemaDoc = go emptyComments
                         Just d -> blueOptional <+> ", default:" <+> pretty d
                     where blueOptional = annotate Blue "optional"
                in vsep
-                    [ keyDoc <> ":" <+> mkComment requiredDoc,
+                    [ annotate White keyDoc <> ":" <+> mkComment requiredDoc,
```
Comment (NorfairKing) [inline yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:131]: This is not the only keyDoc, see the `<key>` stuff below.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 7 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:116]: In the top-level case it looks weird (although correct I think?) but if you check a nested object, I think that's an improvement. If you check `Data.Yaml.Value`, this is the correct YAML data type.

**Comment by mohoff** [comment]
> Also please make sure the ci passes by running `nix-shell` in the root and then running `pre-commit run -a`.

I don't have a nix setup nor am I familiar with these commands. Can you please elaborate - a CONTRIBUTING.md would be sweet - generally and to help with the CI issues of this MR

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 8 above)
Comment (mohoff) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:131]: ah, I see. Done

**Comment by NorfairKing** [comment]
@mohoff Don't worry then, I'll get CI working. Just ping me here when you're done.

**Comment by mohoff** [comment]
@NorfairKing Done from my side. I left the discussions unresolved.

File: yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs (hunk 7 above)
Comment (NorfairKing) [reply yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:116]: Ah yes, your'e right. I'll add a special case for the top level like this and revert the normal case:
```
            ObjectSchema t AnySchema -> e "<object>" (addMComment cs t)
            ObjectSchema t s -> ge s
```

**Comment by NorfairKing** [comment]
Merged manually, thanks!!
